### PR TITLE
Check MACH_PORT_VALID on return from pthread_mach_thread_np

### PR DESCRIPTION
**Pull request by indragiek**
## :scroll: Description

`pthread_mach_thread_np` can return invalid values so check for those and return nullptr from `ThreadHandle`. That means changing the call sites for `ThreadHandle::current()` to check for a null pointer value.

## :bulb: Motivation and Context

We don't have a solid repro for https://github.com/getsentry/sentry-cocoa/issues/3354 but this is one of the potential causes. In any case it doesn't hurt to add the check.

## :green_heart: How did you test it?

Run tests

## :pencil: Checklist

You have to check all boxes before merging:

- [X] I reviewed the submitted code.
- [ ] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: Sources/Sentry/include/SentryThreadHandle.hpp [hunk 1]
```diff
@@ -69,6 +71,8 @@ namespace profiling {
          * the threads in the current process, excluding the current thread (the
          * thread that this function is being called on), and the second element
          * is a handle to the current thread.
+         *
+         * Returns ({}, nullptr) if the current thread could not be retrieved.
          */
         static std::pair<std::vector<std::unique_ptr<ThreadHandle>>, std::unique_ptr<ThreadHandle>>
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentryThreadHandle.hpp:77]: `h`: If I'm not mistaken, we have to check for a potential `nullptr` of this method here https://github.com/getsentry/sentry-cocoa/blob/48e8c2e10e5b5e103fd28ed34b54ca7a9a7645ca/Sources/Sentry/SentryBacktrace.cpp#L159-L160 or does `backtrace` handle `nullptr` correctly?

**Review by philipphofmann — COMMENTED** [review]
@indragiek, good catch 👍 . CI is broken, and it would be great to add or change some tests for these changes if possible.

File: Sources/Sentry/SentryThreadHandle.cpp [hunk 2]
```diff
@@ -74,6 +77,10 @@ namespace profiling {
         mach_msg_type_number_t count;
         thread_act_array_t list;
         auto current = ThreadHandle::current();
+        if (current == nullptr) {
+            return std::make_pair(threads, nullptr);
```
[the remark this answers is not in the document]
Comment (armcknight) [reply Sources/Sentry/SentryThreadHandle.cpp:81]: I just went back to read about [std::move](https://en.cppreference.com/w/cpp/utility/move) again and still don't understand what it's meant for XD Please let us know @indragiek .

File: Sources/Sentry/include/SentryThreadHandle.hpp (hunk 1 above)
Comment (armcknight) [reply Sources/Sentry/include/SentryThreadHandle.hpp:77]: Good catch, doesn't look like there's a check for nullptr before this dereference.

**Review by armcknight — COMMENTED** [review]
Reading the logic again in SentryBacktrace.cpp which contains the callsite for the function we've modified is a bit unclear to me, given the name `allExcludingCurrent`. Reading it starting at the loop, it looks like we're going to do work on all the other threads except the current one, which is true, but we do still have to retain the current thread in order to know what thread we're running on, and the callsite tells me we threw out that info when we actually didn't. 

TL;DR I would propose we rename `allExcludingCurrent` to `currentAndOtherThreads`.

File: Sources/Sentry/SentryThreadHandle.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (armcknight) [reply Sources/Sentry/SentryThreadHandle.cpp:81]: I think this might also be causing this error which is breaking CI:

> /Applications/Xcode15.1/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator17.2.sdk/usr/include/c++/v1/__memory/allocator.h:168:28 Call to implicitly-deleted copy constructor of 'std::unique_ptr<sentry::profiling::ThreadHandle>'

[the remark this answers is not in the document]
Comment (armcknight) [reply Sources/Sentry/SentryThreadHandle.cpp:81]: Back on the question about `std::move`, I just came across this SO answer that mentions it wrt `std::unique_ptr`: https://stackoverflow.com/a/6876833, that mentions using std::move when you need to create a subsequent reference to an object managed by a `unique_ptr`, which can only ever have one reference to it. Is there another use for std::move, or is this the gist of it @indragiek ? In that case, I think we don't need std::move here, because `threads` is a vector of `unique_ptr`, not a `unique_ptr` itself. Looking for confirmation on that though.

[the remark this answers is not in the document]
Comment (indragiek) [reply Sources/Sentry/SentryThreadHandle.cpp:81]: It's not needed here:`std::move` on the vector would move the vector into the `std::pair` instead of making a copy of the vector and then storing the copy in the pair, but since the vector is empty here it doesn't make any difference

On the usage of `std::move` line 98 the vector has elements in it so we do avoid copying a vector by moving it instead

[the remark this answers is not in the document]
Comment (armcknight) [reply Sources/Sentry/SentryThreadHandle.cpp:81]: Makes sense, thanks!

File: Sources/Sentry/include/SentryThreadHandle.hpp (hunk 1 above)
Comment (armcknight) [reply Sources/Sentry/include/SentryThreadHandle.hpp:77]: I'm trying to figure out what is the right course of action if that is null...

The `pair` that is returned by `allThreadsExcludingCurrent` contains in the first spot the list of all threads excluding the current thread, and in the second spot, the current thread. So, `pair.second == nullptr` means we couldn't retrieve a handle for the current thread, due to either `pthread_mach_thread_np` returning `nullptr`, or its return value failing the check in `MACH_PORT_VALID`.

The only place `pair.second`, the "calling thread", is used in `backtrace(...)` is to check if the "target thread" is the same as the current (or "calling") thread (this branch is coded as `UNLIKELY`). And in that case, we set the value of the current stack walk frame pointer to `__builtin_frame_address(0)`. After that, we check if that pointer is valid in the current stack bounds (also coded `UNLIKELY`), and if not, we bail early without continuing up the stack, and the stack will effectively be discarded because we don't collect partial stacks.

So if `pair.second` is `nullptr`, should we pass in the current "target" thread instead, which leads to the stack pointer being set to `__builtin_frame_address(0)`? The other alternative is to not walk any of the threads but that doesn't seem right, as there's still data there to collect.

Comment (armcknight) [reply Sources/Sentry/include/SentryThreadHandle.hpp:77]: Ah, I just noticed that if `allExcludingCurrent` can't get the current thread handle, the pair it returns has null for the current thread (second) and an _empty_ vector in the first part. So, no backtraces will be enumerated in this case. I think that clears up what to do. Just working on a test for this now.

**Review by philipphofmann — COMMENTED** [review]
Once we have some tests, I will approve this PR. The rest LGTM.

File: Sources/Sentry/include/SentryThreadHandle.hpp (hunk 1 above)
Comment (armcknight) [reply Sources/Sentry/include/SentryThreadHandle.hpp]: Fixed in https://github.com/getsentry/sentry-cocoa/pull/4076

**Comment by philipphofmann** [comment]
@armcknight, I guess after resolving the conflicts, I can review this?

**Comment by philipphofmann** [comment]
I'm going to close this stale PR for now. We can reopen it once we get back to it.
